src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs

```diff
@@ -250,28 +256,57 @@ private static void SetFatalResponse(RequestContext context, int status)
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
-            _stopping = true;
-            // Wait for active requests to drain
-            if (_outstandingRequests > 0)
+            void RegisterCancelation()
             {
-                LogHelper.LogInfo(_logger, "Stopping, waiting for " + _outstandingRequests + " request(s) to drain.");
-
-                var waitForStop = new TaskCompletionSource<object>();
                 cancellationToken.Register(() =>
                 {
-                    LogHelper.LogInfo(_logger, "Timed out, terminating " + _outstandingRequests + " request(s).");
-                    waitForStop.TrySetResult(0);
+                    if (Interlocked.Exchange(ref _shutdownSignalCompleted, 1) == 0)
+                    {
+                        LogHelper.LogInfo(_logger, "Canceled, terminating " + _outstandingRequests + " request(s).");
+                        _shutdownSignal.TrySetResult(null);
```

Comment:
This only differs from the current code by having a single task instead of newing up a TCS per call. I couldn't find any code that actually aborts the requests. @Tratcher can you shed some light here?